File: Sources/Server/Source/Entity/Player/PlayerPetalOrbit.ts
import { Entity, EntityMixinConstructor, EntityMixinTemplate, onUpdateTick } from "../Entity";
import { WavePool, UPDATE_ENTITIES_FPS } from "../../Wave/WavePool";
import { BasePlayer, Player } from "./Player";
import { isLivingSlot, PetalData, PetalStat } from "../Mob/Petal/Petal";
import { isPetal } from "../../Utils/common";
import { PetalType } from "../../../../Shared/EntityType";
import { PETAL_PROFILES } from "../../../../Shared/Entity/Mob/Petal/petalProfiles";
import { decodeMood, Mood } from "../../../../Shared/mood";
import { findNearestEntity } from "../Mob/MobAggressivePursuit";
import { MobInstance } from "../Mob/Mob";

export const MAX_CLUSTER_AMOUNT = 5;

const TAU = Math.PI * 2;

const HISTORY_SIZE = 10;

const PRECALC_SIZE = 360;
const lazyCosTable = new Float32Array(PRECALC_SIZE);
const lazySinTable = new Float32Array(PRECALC_SIZE);
for (let i = 0; i < PRECALC_SIZE; i++) {
    const angle = (i * TAU) / PRECALC_SIZE;
    lazyCosTable[i] = Math.cos(angle);
    lazySinTable[i] = Math.sin(angle);
}

const calcTableIndex = (i: number) => (((i % TAU + TAU) % TAU) * PRECALC_SIZE / TAU) | 0;

export const UNMOODABLE_PETALS: Set<PetalType> = new Set([
    PetalType.BEETLE_EGG,
]);

export function PlayerPetalOrbit<T extends EntityMixinConstructor<BasePlayer>>(Base: T) {
    return class MixedBase extends Base implements EntityMixinTemplate {
        private static readonly DEFAULT_ROTATE_SPEED = 2.5;

        private static readonly FRICTION_COEFFICIENT = 0.75;

        private static readonly SPRING_STRENGTH = 0.15;

        private static readonly PETAL_CLUSTER_RADIUS = 8;

        private static readonly SPIN_COEFFICIENT = 0.1;
        private static readonly SPIN_ANGLE_COEFFICIENT = 5;
        private static readonly SPIN_SIN_LOWER = 0.75;

        private rotation = 0;
        private historyIndex = 0;
        private historyX: Float32Array = new Float32Array(HISTORY_SIZE);
        private historyY: Float32Array = new Float32Array(HISTORY_SIZE);
        private petalRadii: Float32Array;
        private petalVelocities: Float32Array;
        private petalSpins: Float32Array[];

        [onUpdateTick](poolThis: WavePool): void {
            super[onUpdateTick](poolThis);

            this.historyX[this.historyIndex] = this.x;
            this.historyY[this.historyIndex] = this.y;
            const historyTargetIndex = (this.historyIndex + 8) % HISTORY_SIZE;
            this.historyIndex = (this.historyIndex + 1) % HISTORY_SIZE;

            const surface = this.slots.surface;
            const totalPetals = surface.length;

            if (!this.petalRadii || this.petalRadii.length !== totalPetals) {
                this.petalRadii = new Float32Array(totalPetals).fill(40);
                this.petalVelocities = new Float32Array(totalPetals).fill(0);
                this.petalSpins = new Array(totalPetals).fill(null).map(() => new Float32Array(MAX_CLUSTER_AMOUNT).fill(0));
            }

            const { 0: isAngry, 1: isSad } = decodeMood(this.mood);

            const numYinYang = surface.reduce(
                (acc, curr) => acc + (
                    curr && isLivingSlot(curr) && curr.length > 0 && curr[0].type === PetalType.YIN_YANG ? 1 : 0
                ),
                0
            );

            // Basically, every 2 yin yangs adds 1 ring
            // Yin yang changes the number of petals per ring by diving it by floor(num yin yang / 2) + 1

            const numRings = Math.floor(numYinYang / 2) + 1;

            const clockwise = numYinYang % 2;

            let realLength = 0;
            for (let i = 0; i < totalPetals; i++) {
                const petals = surface[i];
                if (!petals || !isLivingSlot(petals)) continue;

                const firstPetal = petals[0];
                const profile = PETAL_PROFILES[firstPetal.type][firstPetal.rarity];
                realLength += profile.isCluster ? 1 : petals.length;
            }

            realLength = Math.ceil(realLength / numRings);

            let currentAngleIndex = 0;

            const targetX = this.historyX[historyTargetIndex];
            const targetY = this.historyY[historyTargetIndex];

            let totalSpeed = MixedBase.DEFAULT_ROTATE_SPEED;

            const spinRotationDelta =
                this.calculateRotationDelta(MixedBase.DEFAULT_ROTATE_SPEED, clockwise) *
                MixedBase.SPIN_ANGLE_COEFFICIENT;

            for (let i = 0; i < totalPetals; i++) {
                const petals = surface[i];
                if (!petals || !isLivingSlot(petals)) continue;

                const firstPetal = petals[0];
                if (!isPetal(firstPetal.type)) continue;

                const profile: PetalData = PETAL_PROFILES[firstPetal.type];
                const rarityProfile: PetalStat = profile[firstPetal.rarity];

                // Add faster speed
                if (firstPetal.type === PetalType.FASTER) {
                    totalSpeed += rarityProfile.rad;
                }

                let targetRadius = UNMOODABLE_PETALS.has(firstPetal.type) ? 40 :
                    isAngry ? 80 :
                        isSad ? 25 :
                            40;

                // 25 is FLOWER_ARC_RADIUS
                targetRadius += ((this.size / Player.BASE_SIZE) - 1) * 25;

                const springForce = (targetRadius - this.petalRadii[i]) * MixedBase.SPRING_STRENGTH;
                this.petalVelocities[i] = this.petalVelocities[i] * MixedBase.FRICTION_COEFFICIENT + springForce;
                this.petalRadii[i] += this.petalVelocities[i];

                const ringIndex = Math.floor(currentAngleIndex / realLength);
                const rad = this.petalRadii[i] * (1 + (ringIndex * 0.5));

                const multipliedRotation = this.rotation * (1 + ((ringIndex - (numRings - 1)) * 0.1));

                const baseAngle = TAU * (currentAngleIndex % realLength) / realLength + multipliedRotation;
                currentAngleIndex++;

                if (
                    rarityProfile.isCluster &&
                    petals.length > 1
                ) {
                    const angleIndex = calcTableIndex(baseAngle);
                    const slotBaseX = targetX + lazyCosTable[angleIndex] * rad;
                    const slotBaseY = targetY + lazySinTable[angleIndex] * rad;

                    for (let j = 0; j < petals.length; j++) {
                        const petal = petals[j];

                        // Bit faster than orbit
                        const petalAngle = TAU * j / petals.length + multipliedRotation * 1.1;

                        this.doPetalOrbit(
                            petal,
                            slotBaseX,
                            slotBaseY,
                            MixedBase.PETAL_CLUSTER_RADIUS,
                            petalAngle,
                        );

                        const petalSpin = this.petalSpins[i];

                        if (
                            this.doPetalSpin(
                                poolThis,
                                petal,
                                petalSpin[j],
                            )
                        ) {
                            petalSpin[j] += spinRotationDelta;
                        };
                    }
                } else {
                    const petal = petals[0];

                    this.doPetalOrbit(
                        petal,
                        targetX,
                        targetY,
                        rad,
                        baseAngle
                    );

                    const petalSpin = this.petalSpins[i];

                    if (
                        this.doPetalSpin(
                            poolThis,
                            petal,
                            petalSpin[0],
                        )
                    ) {
                        petalSpin[0] += spinRotationDelta;
                    };
                }
            }

            const rotationDelta = this.calculateRotationDelta(totalSpeed, clockwise);
            this.rotation += rotationDelta;

            if (Math.abs(this.rotation) > Number.MAX_SAFE_INTEGER) {
                this.rotation = this.rotation % TAU;
            }
        }

        private calculateRotationDelta(
            totalSpeed: number,
            clockwise: number,
        ): number {
            return (totalSpeed * (clockwise ? -1 : 1)) / UPDATE_ENTITIES_FPS
        }

        private doPetalOrbit(
            petal: MobInstance,
            targetX: number,
            targetY: number,
            holdingRadius: number,
            angle: number
        ) {
            const angleIndex = calcTableIndex(angle);

            const chaseX = targetX + lazyCosTable[angleIndex] * holdingRadius;
            const chaseY = targetY + lazySinTable[angleIndex] * holdingRadius;

            const diffX = chaseX - petal.x;
            const diffY = chaseY - petal.y;

            const ACCELERATION = 0.25;

            const { petalVelocity: velocity } = petal;

            if (!velocity) return;

            velocity[0] += ACCELERATION * diffX;
            velocity[1] += ACCELERATION * diffY;

            velocity[0] *= MixedBase.FRICTION_COEFFICIENT;
            velocity[1] *= MixedBase.FRICTION_COEFFICIENT;

            petal.x += velocity[0];
            petal.y += velocity[1];
        }

        /**
         * Do petal spin within petal.
         * 
         * @returns - Is petal spinned or not.
         */
        private doPetalSpin(
            poolThis: WavePool,
            petal: MobInstance,
            rotation: number,
        ): boolean {
            const mobToSpin = findNearestEntity(
                petal,
                poolThis.getAllMobs()
                    .filter(p =>
                        !(
                            p.id === petal.id ||
                            isPetal(p.type) ||
                            p.petMaster
                        )
                    )
                    .filter(entity => {
                        const distance = Math.hypot(
                            entity.x - petal.x,
                            entity.y - petal.y
                        );

                        return distance <= (entity.size * (1 + MixedBase.SPIN_COEFFICIENT));
                    }),
            );

            petal.petalSpinningMob = !!mobToSpin;

            if (petal.petalSpinningMob) {
                const spiralRadius =
                    mobToSpin.size * (1 - MixedBase.SPIN_COEFFICIENT) *
                    (MixedBase.SPIN_SIN_LOWER + (Math.sin(rotation % TAU) + 1) / 2 * (1 - MixedBase.SPIN_SIN_LOWER));

                const spinAngleIndex = calcTableIndex(rotation);

                petal.x = mobToSpin.x + lazyCosTable[spinAngleIndex] * spiralRadius;
                petal.y = mobToSpin.y + lazySinTable[spinAngleIndex] * spiralRadius;
            }

            return petal.petalSpinningMob;
        }

        dispose(): void {
            if (super.dispose) {
                super.dispose();
            }

            this.historyX = null;
            this.historyY = null;
            this.petalRadii = null;
            this.petalVelocities = null;

            for (let i = 0; i < this.petalSpins.length; i++) {
                // Dont forgot to release object items too
                this.petalSpins[i] = null;
            }
            this.petalSpins = null;
        }
    };
}

File: Sources/server/internal/wave/player_petal_orbit.go
package wave

import (
	"math"
	"slices"
	"time"

	"flooooio/internal/wave/florr/native"

	"github.com/chewxy/math32"

	"github.com/youdie323323/go-spatial-hash"
)

const (
	OrbitHistorySize = 10

	precalcSize = 360
)

var (
	lazyCosTable = make([]float32, precalcSize)
	lazySinTable = make([]float32, precalcSize)
)

func init() {
	for i := range precalcSize {
		angle := (float32(i) * Tau32) / precalcSize

		lazyCosTable[i] = math32.Cos(angle)
		lazySinTable[i] = math32.Sin(angle)
	}
}

var UnmoodablePetalTypes = slices.Concat(UsablePetalTypes, []native.PetalType{
	native.PetalTypeMagnet,
})

// PlayerPetalOrbit class constants
const (
	defaultRotateSpeed        = 2.5
	radiusSpringStrength      = 0.4
	radiusFriction            = 0.1
	petalVelocityAcceleration = 0.1
	petalClusterRadius        = 6.0
	spingInterpolationSpeed   = 0.6
	spinNearestSizeCoef       = 1.075
	spinAngleCoef             = 10.0
)

const (
	DefaultMoodRadius = 40
	SadMoodRadius     = 25
	AngryMoodRadius   = 80
)

func calculateTableIndex(angle float32) int {
	return int(math32.Mod(math32.Mod(angle, Tau32)+Tau32, Tau32) * 180 / math32.Pi)
}

func calculateRotationDelta(
	speed float32,

	clockwise float32,
) float32 {
	return (speed * clockwise) * DeltaT
}

func orbitPetal(
	petal *Petal,

	targetX float32,
	targetY float32,

	radius float32,

	angle float32,
) {
	angleIndex := calculateTableIndex(angle)

	chaseX := targetX + lazyCosTable[angleIndex]*radius
	chaseY := targetY + lazySinTable[angleIndex]*radius

	diffX := chaseX - petal.X
	diffY := chaseY - petal.Y

	petal.Velocity[0] += petalVelocityAcceleration * diffX
	petal.Velocity[1] += petalVelocityAcceleration * diffY
}

// spinPetal do petal spin on mob.
func spinPetal(
	wp *Pool,

	petal *Petal,

	spins [][]float32,
	i int,
	j int,
) {
	mobToSpin, ok := FindNearestEntity(
		petal,
		spatial_hash.ToNodeSlice(wp.FilterMobsWithCondition(func(m *Mob) bool {
			if !m.IsOrganismEnemy() {
				return false
			}

			spinDetectRad := m.Radius() * spinNearestSizeCoef
			spinDetectRadSq := spinDetectRad * spinDetectRad

			dx := m.X - petal.X
			dy := m.Y - petal.Y

			return (dx*dx + dy*dy) <= spinDetectRadSq
		})),
	).(*Mob)
	if !ok {
		return
	}

	wasSpinning := petal.SpinningOnMob

	petal.SpinningOnMob = mobToSpin != nil

	if petal.SpinningOnMob {
		if !wasSpinning {
			targetAngle := math32.Atan2(
				petal.Y-mobToSpin.Y,
				petal.X-mobToSpin.X,
			)

			angleDiff := targetAngle - spins[i][j]

			angleDiff = math32.Mod(math32.Mod(angleDiff, Tau32)+Tau32, Tau32)
			if angleDiff > math32.Pi {
				angleDiff -= Tau32
			}

			spins[i][j] = math32.Mod(spins[i][j]+angleDiff, Tau32)
		}

		spinAngleIndex := calculateTableIndex(spins[i][j])

		mobToSpinDesiredSize := 1.1 * mobToSpin.Radius()

		targetX := mobToSpin.X + lazyCosTable[spinAngleIndex]*mobToSpinDesiredSize
		targetY := mobToSpin.Y + lazySinTable[spinAngleIndex]*mobToSpinDesiredSize

		petal.X += (targetX - petal.X) * spingInterpolationSpeed
		petal.Y += (targetY - petal.Y) * spingInterpolationSpeed
	}
}

func (p *Player) PlayerPetalOrbit(wp *Pool, now time.Time) {
	p.OrbitHistoryX[p.OrbitHistoryIndex] = p.X
	p.OrbitHistoryY[p.OrbitHistoryIndex] = p.Y

	historyTargetIndex := (p.OrbitHistoryIndex + 6) % OrbitHistorySize

	targetX := p.OrbitHistoryX[historyTargetIndex]
	targetY := p.OrbitHistoryY[historyTargetIndex]

	p.OrbitHistoryIndex = (p.OrbitHistoryIndex + 1) % OrbitHistorySize

	surface := p.Slots.Surface
	surfaceLen := len(surface)

	if p.OrbitPetalRadii == nil {
		p.OrbitPetalRadii = make([]float32, surfaceLen)

		for i := range p.OrbitPetalRadii {
			p.OrbitPetalRadii[i] = 40
		}

		p.OrbitRadiusVelocities = make([]float32, surfaceLen)

		p.OrbitPetalSpins = make([][]float32, surfaceLen)

		for i := range p.OrbitPetalSpins {
			p.OrbitPetalSpins[i] = make([]float32, PetalMaxClusterAmount)
		}
	}

	isAngry := p.Mood.IsSet(native.MoodAngry)
	isSad := p.Mood.IsSet(native.MoodSad)

	var numYinYang float32 = 0.

	for _, petals := range surface {
		if petals == nil {
			continue
		}

		for _, petal := range petals {
			if petal != nil && petal.Type == native.PetalTypeYinYang {
				numYinYang++

				break
			}
		}
	}

	// Basically, every 2 yin yangs adds 1 ring
	// Yin yang changes the number of petals per ring by diving it by floor(num yin yang / 2) + 1

	numRings := math32.Floor(numYinYang/2) + 1

	var clockwise float32

	if math32.Mod(numYinYang, 2) == 0 {
		clockwise = 1
	} else {
		clockwise = -1
	}

	realLength := float32(len(surface))

	realLength = math32.Ceil(realLength / numRings)

	var currentAngleIndex float32 = 0.

	var totalSpeed float32 = defaultRotateSpeed

	spinRotationDelta := calculateRotationDelta(defaultRotateSpeed*spinAngleCoef, clockwise)

	var targetRadius float32

	switch true {
	case isAngry:
		targetRadius = AngryMoodRadius

	case isSad:
		targetRadius = SadMoodRadius

	default:
		targetRadius = DefaultMoodRadius
	}

	targetRadius += ((p.Size / PlayerSize) - 1) * PlayerCollision.Radius

	nowMilli := float64(now.UnixMilli())

	for i, petals := range surface {
		if petals == nil {
			continue
		}

		firstPetal := petals[0]
		if firstPetal == nil {
			continue
		}

		firstPetalStat := native.PetalProfiles[firstPetal.Type].StatFromRarity(firstPetal.Rarity)

		if firstPetal.Type == native.PetalTypeFaster {
			rad, ok := firstPetalStat.Extra["rad"]
			if ok {
				totalSpeed += rad
			}
		}

		var springForce float32

		switch true {
		case slices.Contains(UnmoodablePetalTypes, firstPetal.Type):
			springForce = (DefaultMoodRadius - p.OrbitPetalRadii[i]) * radiusSpringStrength

		case isAngry && firstPetal.Type == native.PetalTypeWing:
			wingAddition := float32(130 * (math.Sin(math.Mod((nowMilli+float64(firstPetal.Id))/200, Tau)) + 1))

			springForce = (targetRadius + wingAddition - p.OrbitPetalRadii[i]) * radiusSpringStrength

		default:
			springForce = (targetRadius - p.OrbitPetalRadii[i]) * radiusSpringStrength
		}

		p.OrbitRadiusVelocities[i] = p.OrbitRadiusVelocities[i]*radiusFriction + springForce
		p.OrbitPetalRadii[i] += p.OrbitRadiusVelocities[i]

		ringIndex := math32.Floor(currentAngleIndex / realLength)
		radius := p.OrbitPetalRadii[i] * ((0.5 * ringIndex) + 1)

		multipliedRotation := p.OrbitRotation * (1 + ((ringIndex - (numRings - 1)) * 0.1))

		baseAngle := Tau32*math32.Mod(currentAngleIndex, realLength)/realLength + multipliedRotation
		currentAngleIndex++

		if IsClusterPetal(petals) {
			angleIndex := calculateTableIndex(baseAngle)

			slotBaseX := targetX + lazyCosTable[angleIndex]*radius
			slotBaseY := targetY + lazySinTable[angleIndex]*radius

			for j, petal := range petals {
				if petal == nil {
					continue
				}

				if petal.Detached {
					continue
				}

				// Bit faster than normal orbit
				petalAngle := Tau32*float32(j)/float32(len(petals)) + multipliedRotation*1.3

				orbitPetal(
					petal,

					slotBaseX,
					slotBaseY,

					petalClusterRadius,

					petalAngle,
				)

				// Usable petal type never spins on mob
				if slices.Contains(UsablePetalTypes, petal.Type) {
					continue
				}

				spins := p.OrbitPetalSpins
				if len(spins) > i {
					spin := spins[i]
					if len(spin) > j {
						spinPetal(
							wp,

							petal,

							spins,
							i,
							j,
						)

						if petal.SpinningOnMob {
							spin[j] += spinRotationDelta
						}
					}
				}
			}
		} else {
			petal := petals[0]

			if petal == nil {
				continue
			}

			if petal.Detached {
				continue
			}

			orbitPetal(
				petal,

				targetX,
				targetY,

				radius,

				baseAngle,
			)

			// Usable petal type never spins on mob
			if slices.Contains(UsablePetalTypes, petal.Type) {
				continue
			}

			spins := p.OrbitPetalSpins
			if len(spins) > i {
				spin := spins[i]
				if len(spin) > 0 {
					spinPetal(
						wp,

						petal,

						spins,
						i,
						0,
					)

					if petal.SpinningOnMob {
						spin[0] += spinRotationDelta
					}
				}
			}
		}
	}

	p.OrbitRotation += calculateRotationDelta(totalSpeed, clockwise)

	// Limit in the tau
	if math32.Abs(p.OrbitRotation) >= math32.MaxFloat32 {
		p.OrbitRotation = math32.Mod(p.OrbitRotation, Tau32)
	}
}
